R/pivot_quantiles.R

```diff
@@ -0,0 +1,160 @@
+#' Turn a vector of quantile distributions into a list-col
+#'
+#' @param x a `distribution` containing `dist_quantiles`
+#'
+#' @return a list-col
+#' @export
+#'
+#' @examples
+#' edf <- case_death_rate_subset[1:3, ]
+#' edf$q <- dist_quantiles(list(1:5, 2:4, 3:10), list(1:5 / 6, 2:4 / 5, 3:10 / 11))
+#'
+#' edf_nested <- edf %>% dplyr::mutate(q = nested_quantiles(q))
+#' edf_nested %>% tidyr::unnest(q)
+nested_quantiles <- function(x) {
+  stopifnot(is_dist_quantiles(x))
+  distributional:::dist_apply(x, .f = function(z) {
+    tibble::as_tibble(vec_data(z)) %>%
+      dplyr::mutate(dplyr::across(tidyselect::everything(), as.double)) %>%
+      list_of()
+  })
+}
+
+
+#' Pivot columns containing `dist_quantile` longer
+#'
+#' Selected columns that contains `dist_quantiles` will be "lengthened" with
+#' the "taus" (quantile) serving as 1 column and the values as another. If
+#' multiple columns are selected, these will be prefixed the the column name.
+#'
+#' @param .data A data frame, or a data frame extension such as a tibble or
+#'   epi_df.
+#' @param ... <[`tidy-select`][dplyr::dplyr_tidy_select]> One or more unquoted
+#'   expressions separated by commas. Variable names can be used as if they
+#'   were positions in the data frame, so expressions like `x:y` can
+#'   be used to select a range of variables.
+#' @param .ignore_length_check If multiple columns are selected, as long as
+#'   each row has contains the same number of quantiles, the result will be
+#'   reasonable. But if, for example, `var1[1]` has 5 quantiles while `var2[1]`
+#'   has 7, then the only option would be to recycle everything, creating a
+#'   _very_ long result. By default, this would throw an error. But if this is
+#'   really the goal, then the error can be bypassed by setting this argument
+#'   to `TRUE`. The first selected column will vary fastest.
+#'
+#' @return An object of the same class as `.data`.
+#' @export
+#'
+#' @examples
+#' d1 <- c(dist_quantiles(1:3, 1:3 / 4), dist_quantiles(2:4, 1:3 / 4))
+#' d2 <- c(dist_quantiles(2:4, 2:4 / 5), dist_quantiles(3:5, 2:4 / 5))
+#' tib <- tibble::tibble(g = c("a", "b"), d1 = d1, d2 = d2)
+#'
+#' pivot_quantiles_longer(tib, "d1")
```

Comment:
note: Want to verify that this is the expected behavior:

![Screenshot from 2023-10-04 17-15-56](https://github.com/cmu-delphi/epipredict/assets/42820733/956d195f-ca7d-4d31-a789-48d4453fdda1)

I see that the `q` and `tau` columns are inserted at the spot where the old column used to be, instead of always ending up at the end, for example.

Reply:
Yes, this is expected (matches `unnest()`).